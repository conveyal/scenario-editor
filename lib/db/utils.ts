import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollectionResults(obj: unknown) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollectionResults)
  if (typeof obj === 'object') return mapValues(obj, serializeCollectionResults)
  return obj
}
