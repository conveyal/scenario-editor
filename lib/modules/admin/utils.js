import {connect} from 'react-redux'
import {createStructuredSelector} from 'reselect'

import * as actions from './actions'
import * as select from './selectors'

export function fullyConnect(Component) {
  return connect(createStructuredSelector(select), actions)(Component)
}

const add0 = n => (n >= 10 ? n : `0${n}`)

export function msToDuration(ms) {
  const durationS = Math.floor((Date.now() - ms) / 1000)
  const s = durationS % 60
  const m = Math.floor((durationS / 60) % 60)
  const h = Math.floor((durationS / 60 / 60) % 60)

  return `${add0(h)}:${add0(m)}:${add0(s)}`
}

/**
 * Example commit string: "v3.4.0-118-gbe7199c". Take the last section after
 * splitting on the hyphens and remove the "g".
 */
const githubUrl = 'https://github.com/conveyal/r5/commits'
export function createR5Url(workerCommit) {
  // not a release version
  if (workerCommit.indexOf('-') !== -1) {
    const hash = workerCommit.split('-')[2].slice(1) //
    return `${githubUrl}/${hash}`
  }
  return `${githubUrl}/${workerCommit}`
}

export function createWorkerUrl(instanceId, region) {
  return `https://${region}.console.aws.amazon.com/ec2/v2/home?region=${region}#Instances:instanceId=${instanceId};sort=tag:Name`
}

const C1 = '#2389c9'
const C2 = '#36c923'
const C3 = '#aab4bb'

/**
 * Create a progress bar with a linear-gradient.
 */
export function createLinearGradientForJob(c, d, t) {
  const p1 = (c / t) * 100
  const p2 = (d / t) * 100
  return `linear-gradient(to right,
    ${C1} 0%,
    ${C1} ${p1}%,
    ${C2} ${p1}%,
    ${C2} ${p2}%,
    ${C3} ${p2}%,
    ${C3} 100%)`
}
