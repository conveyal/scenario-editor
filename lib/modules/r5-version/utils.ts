import {
  MINIMUM_R5_VERSION,
  VERSION_PARSE_REGEX,
  OLD_VERSION_PARSE_REGEX
} from './constants'

/**
 * Convert an r5 version to a number for comparing
 * NB Assumes no major, minor, patch, or commit version exceeds 100
 */
export function versionToNumber(version?: string): number {
  const versionString = version || ''
  if (versionString.match(/\./g) && versionString.match(/\./g).length == 1) {
    const versionMatch = versionString.match(VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [, major = '0', minor = '0', commit = '0'] = versionMatch

      return (
        parseInt(major) * 10000 + parseInt(minor) * 100 + parseInt(commit) / 100
      )
    }

    return 0
  } else {
    const versionMatch = versionString.match(OLD_VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [
        ,
        major = '0',
        minor = '0',
        patch = '0',
        commit = '0'
      ] = versionMatch

      return (
        parseInt(major) * 10000 +
        parseInt(minor) * 100 +
        parseInt(patch) +
        parseInt(commit) / 100
      )
    }

    return 0
  }
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = (version || '').search(/-/g) !== -1
  return isTestVersion || workerVersionInRange(version, min, max)
}
