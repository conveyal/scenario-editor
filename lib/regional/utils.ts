function getDefaultCutoff(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.cutoffsMinutes))
    return a.cutoffsMinutes[Math.floor(a.cutoffsMinutes.length / 2)]
  return a.cutoffMinutes
}

function getDefaultDestinationPointSet(a: CL.RegionalAnalysis): string | void {
  const ids = a.destinationPointSetIds
  if (Array.isArray(ids)) return ids[0]
}

function getDefaultPercentile(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.travelTimePercentiles))
    return a.travelTimePercentiles[
      Math.floor(a.travelTimePercentiles.length / 2)
    ]
  return a.travelTimePercentile
}

/**
 * Choose
 */
export function getDefaultVariants(analysis: CL.RegionalAnalysis) {
  return {
    cutoff: getDefaultCutoff(analysis),
    destinationPointSetId: getDefaultDestinationPointSet(analysis),
    percentile: getDefaultPercentile(analysis)
  }
}

export function getComparisonVariants(comparison: CL.RegionalAnalysis) {
  return {
    comparisonCutoff: getDefaultCutoff(comparison),
    comparisonDestinationPointSetId: getDefaultDestinationPointSet(comparison),
    comparisonPercentile: getDefaultPercentile(comparison)
  }
}

/**
 * Don't compare incomparable analyses.
 */
export function getValidComparisonAnalyses(
  allAnalyses: CL.RegionalAnalysis[],
  activeJobs: CL.RegionalJob[],
  analysis: CL.RegionalAnalysis
) {
  return allAnalyses.filter(
    (c) =>
      activeJobs.findIndex((job) => job.jobId === c._id) === -1 &&
      c.request.originPointSetKey == null &&
      c.zoom === analysis.zoom &&
      c.width === analysis.width &&
      c.height === analysis.height &&
      c.north === analysis.north &&
      c.west === analysis.west
  )
}
