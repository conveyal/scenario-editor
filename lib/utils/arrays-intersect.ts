// Do the arrays intersect?
export default function arraysIntersect<T>(a1: T[], a2: T[]) {
  for (const v of a1) {
    if (a2.indexOf(v) > -1) return true
  }
  return false
}
