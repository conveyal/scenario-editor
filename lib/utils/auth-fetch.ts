import {parse} from 'cookie'

import {safeFetch, SafeResponse} from './safe-fetch'

export function createAuthHeaders(user?: CL.User) {
  const headers = {
    Authorization: `bearer ${user?.idToken}`
  }

  // Add the admin access group for administrators if it exists.
  if (user?.accessGroup === process.env.NEXT_PUBLIC_ADMIN_ACCESS_GROUP) {
    const adminTempAccessGroup = parse(document.cookie).adminTempAccessGroup
    if (adminTempAccessGroup?.length > 0) {
      headers['X-Conveyal-Access-Group'] = adminTempAccessGroup
    }
  }
  return headers
}

/**
 * Fetch wrapper that includes authentication. Defaults headers to JSON.
 */
export default function authFetch<T>(
  url: string,
  user: CL.User,
  parser: (res: Response) => Promise<T> = (res) => res.json(),
  options?: RequestInit
): Promise<SafeResponse<T>> {
  return safeFetch(url, parser, {
    mode: 'cors',
    ...options,
    headers: {
      ...createAuthHeaders(user),
      ...options?.headers
    }
  })
}
