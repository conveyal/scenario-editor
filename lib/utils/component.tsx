import {UserProvider} from '@auth0/nextjs-auth0'
import {ThemeProvider, extendTheme} from '@chakra-ui/react'
import {expect} from '@jest/globals'
import enzyme from 'enzyme'
import {Router} from 'next/router'
import {RouterContext} from 'next/dist/next-server/lib/router-context'
import React from 'react'
import {Map} from 'react-leaflet'
import {Provider} from 'react-redux'

import {localUser} from 'lib/user'

import {makeMockStore, mockStores} from './mock-data'

// Next.js uses Math.random to generate unique ids for pages that get captured in snapshots.
global.Math.random = () => 0.5

// Set __NEXT_DATA__ (required by Router)
window.__NEXT_DATA__ = {} as any
const router = new Router('page', {}, 'as-path', {} as any)

const getName = (c) => c.displayName || c.name

const theme = extendTheme({})

export function testComponent(
  TestComponent,
  props = {},
  initialData = mockStores.init
) {
  const store = makeMockStore(initialData)
  const Wrapping = (props) => (
    <UserProvider user={localUser}>
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <RouterContext.Provider value={router}>
            <Map>{props.children}</Map>
          </RouterContext.Provider>
        </ThemeProvider>
      </Provider>
    </UserProvider>
  )

  const mount = () =>
    enzyme.mount(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  const shallow = () =>
    enzyme.shallow(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // For debugging purposes (this is not the original component!)
  const name = `WithContext(${getName(TestComponent)})`

  return {mount, name, router, shallow, store}
}

/**
 * Rename to `wrapComponent`. TODO replace all `testComponent`
 */
export const wrapComponent = testComponent

/**
 * Helper function to create quick snapshot tests for components.
 */
export function testAndSnapshot(C, p, id?: any) {
  test(`Component ${getName(C)} snapshot(mount)`, () => {
    const c = testComponent(C, p, id)
    const t = c.mount()
    expect(t).toMatchSnapshot()
    t.unmount()
  })
}

/**
 * Helper function to mount and unmount a component.
 */
export function testComponentMount(Component, props) {
  test(`Component mount(${getName(Component)})`, () => {
    const c = testComponent(Component, props)
    const t = c.mount()
    t.unmount()
  })
}

/**
 * Test and snapshot map component
 */
export function wrapMapComponent(C, p, id?: any) {
  const MapWrapper = (props) => (
    <Map>
      <C {...props} />
    </Map>
  )
  MapWrapper.displayName = `WithMap(${getName(C)})`
  return testComponent(MapWrapper, p, id)
}

function snapshotTest(wc, type) {
  test(`Component ${getName(wc)} snapshot(${type})`, () => {
    const s = wc[`${type}`]()
    expect(s).toMatchSnapshot()
    s.unmount()
  })
}

/**
 * Shallow snapshot test.
 */
export function shallowSnapshot(wc) {
  snapshotTest(wc, 'shallow')
}

/**
 * Deep snapshot test.
 */
export function deepSnapshot(wc) {
  snapshotTest(wc, 'mount')
}
