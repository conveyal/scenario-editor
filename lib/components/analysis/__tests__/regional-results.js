import {testComponent} from 'lib/utils/component'
import {mockStores, mockGrid, mockRegionalAnalyses} from 'lib/utils/mock-data'
import RegionalResults from '../regional-results'

const props = {
  analysis: mockRegionalAnalyses[0],
  analysisId: mockRegionalAnalyses[0]._id,
  opportunityDatasets: [{name: 'Total jobs', _id: 'Jobs_total'}],
  regionId: 'MOCK_REGION_ID'
}

test('RegionalResults snapshot(mount)', () => {
  const c = testComponent(RegionalResults, props)
  const t = c.mount()
  expect(t).toMatchSnapshot()
  t.unmount()
})

test('RegionalResults with comparison snapshot(mount)', () => {
  const store = {
    ...mockStores.init,
    regionalAnalyses: {
      ...mockStores.init.regionalAnalyses,
      comparisonId: mockRegionalAnalyses[1]._id
    }
  }
  const c = testComponent(RegionalResults, props, store)
  const t = c.mount()
  expect(t).toMatchSnapshot()
  t.unmount()
})
