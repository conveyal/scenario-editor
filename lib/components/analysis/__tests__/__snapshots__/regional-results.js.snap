// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegionalResults snapshot(mount) 1`] = `
<RegionalResults
  analysis={
    Object {
      "_id": "abcd",
      "accessGroup": "accessGroupName",
      "bounds": Object {},
      "bundleId": "5678",
      "complete": true,
      "createdAt": "2017-11-08T12:30:50.133Z",
      "createdBy": "person@gmail.com",
      "cutoffMinutes": 60,
      "deleted": false,
      "grid": "Jobs_total",
      "height": 300,
      "name": "ANALYSIS 1",
      "nonce": "12345",
      "north": 20024,
      "projectId": "1234",
      "request": Object {
        "accessModes": "WALK",
        "bikeSafe": 1,
        "bikeSlope": 1,
        "bikeSpeed": 4.166666666666667,
        "bikeTime": 1,
        "bounds": Object {
          "east": -76,
          "north": 39,
          "south": 38,
          "west": -77,
        },
        "bundleId": "1",
        "date": "2016-01-16",
        "defaultExpanded": true,
        "directModes": "WALK",
        "egressModes": "WALK",
        "fromLat": 12,
        "fromLon": 34,
        "fromTime": 25200,
        "maxBikeTime": 20,
        "maxRides": 4,
        "maxTripDurationMinutes": 60,
        "maxWalkTime": 20,
        "monteCarloDraws": 200,
        "name": "Profile Request Test",
        "projectId": "1",
        "reachabilityThreshold": 0,
        "regionId": "1",
        "toTime": 32400,
        "transitModes": "BUS,TRAM,RAIL,SUBWAY",
        "travelTimePercentile": 50,
        "variantIndex": 0,
        "walkSpeed": 1.3888888888888888,
        "workerVersion": "v4.0.0",
      },
      "status": "DONE",
      "travelTimePercentile": -1,
      "updatedAt": "2017-11-08T12:31:19.602Z",
      "updatedBy": "person@gmail.com",
      "variant": 0,
      "west": 20015,
      "width": 300,
      "workerVersion": "v3.2.0",
      "zoom": 9,
    }
  }
  analysisId="abcd"
  opportunityDatasets={
    Array [
      Object {
        "_id": "Jobs_total",
        "name": "Total jobs",
      },
    ]
  }
  regionId="MOCK_REGION_ID"
>
  <Group
    label="Compare to"
  >
    <div
      className="form-group"
    >
      <label
        className="control-label"
        htmlFor="compare-to-input-0"
      >
        Compare to
      </label>
      <S
        getOptionLabel={[Function]}
        getOptionValue={[Function]}
        isClearable={true}
        onChange={[Function]}
        options={Array []}
      >
        <ReactSelect
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          isClearable={true}
          onChange={[Function]}
          options={Array []}
          styles={
            Object {
              "control": [Function],
              "option": [Function],
            }
          }
        >
          <div
            className="ReactSelectMock"
          />
        </ReactSelect>
      </S>
    </div>
  </Group>
  <Group
    label="Access to"
  >
    <div
      className="form-group"
    >
      <label
        className="control-label"
        htmlFor="access-to-input-1"
      >
        Access to
      </label>
      <P>
        <Box
          as="p"
          mb="10px"
        >
          <p
            className="css-1r2f04i"
          >
            Total jobs within 60 minutes (average instantaneous accessibility)
          </p>
        </Box>
      </P>
      <Legend
        breaks={
          Array [
            0,
            436,
            857,
            1257,
            1649,
            2000,
          ]
        }
        colors={
          Array [
            Object {
              "b": 255,
              "g": 255,
              "opacity": 0,
              "r": 255,
            },
            Rgb {
              "b": 239,
              "g": 219,
              "opacity": 0.42,
              "r": 198,
            },
            Rgb {
              "b": 225,
              "g": 202,
              "opacity": 0.42,
              "r": 158,
            },
            Rgb {
              "b": 214,
              "g": 174,
              "opacity": 0.42,
              "r": 107,
            },
            Rgb {
              "b": 189,
              "g": 130,
              "opacity": 0.42,
              "r": 49,
            },
            Rgb {
              "b": 156,
              "g": 81,
              "opacity": 0.42,
              "r": 8,
            },
          ]
        }
        min={0}
      >
        <div
          className="Legend"
        >
          <div
            className="Legend-Item"
            key="break-0"
            style={
              Object {
                "backgroundColor": "rgba(255, 255, 255, 1)",
                "color": "#000",
              }
            }
          >
            0 (transparent)
          </div>
          <div
            className="Legend-Item"
            key="break-1"
            style={
              Object {
                "backgroundColor": "rgba(198, 219, 239, 1)",
                "color": "#000",
              }
            }
          >
            1 to 436
          </div>
          <div
            className="Legend-Item"
            key="break-2"
            style={
              Object {
                "backgroundColor": "rgba(158, 202, 225, 1)",
                "color": "#000",
              }
            }
          >
            437 to 857
          </div>
          <div
            className="Legend-Item"
            key="break-3"
            style={
              Object {
                "backgroundColor": "rgba(107, 174, 214, 1)",
                "color": "#000",
              }
            }
          >
            858 to 1,257
          </div>
          <div
            className="Legend-Item"
            key="break-4"
            style={
              Object {
                "backgroundColor": "rgba(49, 130, 189, 1)",
                "color": "#fff",
              }
            }
          >
            1,258 to 1,649
          </div>
          <div
            className="Legend-Item"
            key="break-5"
            style={
              Object {
                "backgroundColor": "rgba(8, 81, 156, 1)",
                "color": "#fff",
              }
            }
          >
            1,650 to 2,000
          </div>
        </div>
      </Legend>
    </div>
  </Group>
  <AggregationArea
    regionId="MOCK_REGION_ID"
  >
    <Group
      label="Aggregate results to"
    >
      <div
        className="form-group"
      >
        <label
          className="control-label"
          htmlFor="aggregate-results-to-input-2"
        >
          Aggregate results to
        </label>
        <S
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          isClearable={true}
          isDisabled={false}
          name="aggregateTo"
          onChange={[Function]}
          options={Array []}
        >
          <ReactSelect
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            isClearable={true}
            isDisabled={false}
            name="aggregateTo"
            onChange={[Function]}
            options={Array []}
            styles={
              Object {
                "control": [Function],
                "option": [Function],
              }
            }
          >
            <div
              className="ReactSelectMock"
            />
          </ReactSelect>
        </S>
      </div>
    </Group>
    <Group>
      <div
        className="form-group"
      >
        <Button
          block={true}
          onClick={[Function]}
          size="sm"
          style="info"
        >
          <a
            className="btn btn-info btn-block btn-sm"
            href="#"
            onClick={[Function]}
            tabIndex={0}
          >
            Upload new aggregation area
             
            <Icon
              icon={
                Object {
                  "icon": Array [],
                  "iconName": "chevron-down",
                  "prefix": "fas",
                }
              }
            >
              <FontAwesomeIcon
                border={false}
                className=""
                fixedWidth={true}
                flip={null}
                icon={
                  Object {
                    "icon": Array [],
                    "iconName": "chevron-down",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                swapOpacity={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-chevron-down fa-w-NaN fa-fw "
                  data-icon="chevron-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 undefined undefined"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </Icon>
          </a>
        </Button>
      </div>
    </Group>
  </AggregationArea>
</RegionalResults>
`;

exports[`RegionalResults with comparison snapshot(mount) 1`] = `
<RegionalResults
  analysis={
    Object {
      "_id": "abcd",
      "accessGroup": "accessGroupName",
      "bounds": Object {},
      "bundleId": "5678",
      "complete": true,
      "createdAt": "2017-11-08T12:30:50.133Z",
      "createdBy": "person@gmail.com",
      "cutoffMinutes": 60,
      "deleted": false,
      "grid": "Jobs_total",
      "height": 300,
      "name": "ANALYSIS 1",
      "nonce": "12345",
      "north": 20024,
      "projectId": "1234",
      "request": Object {
        "accessModes": "WALK",
        "bikeSafe": 1,
        "bikeSlope": 1,
        "bikeSpeed": 4.166666666666667,
        "bikeTime": 1,
        "bounds": Object {
          "east": -76,
          "north": 39,
          "south": 38,
          "west": -77,
        },
        "bundleId": "1",
        "date": "2016-01-16",
        "defaultExpanded": true,
        "directModes": "WALK",
        "egressModes": "WALK",
        "fromLat": 12,
        "fromLon": 34,
        "fromTime": 25200,
        "maxBikeTime": 20,
        "maxRides": 4,
        "maxTripDurationMinutes": 60,
        "maxWalkTime": 20,
        "monteCarloDraws": 200,
        "name": "Profile Request Test",
        "projectId": "1",
        "reachabilityThreshold": 0,
        "regionId": "1",
        "toTime": 32400,
        "transitModes": "BUS,TRAM,RAIL,SUBWAY",
        "travelTimePercentile": 50,
        "variantIndex": 0,
        "walkSpeed": 1.3888888888888888,
        "workerVersion": "v4.0.0",
      },
      "status": "DONE",
      "travelTimePercentile": -1,
      "updatedAt": "2017-11-08T12:31:19.602Z",
      "updatedBy": "person@gmail.com",
      "variant": 0,
      "west": 20015,
      "width": 300,
      "workerVersion": "v3.2.0",
      "zoom": 9,
    }
  }
  analysisId="abcd"
  opportunityDatasets={
    Array [
      Object {
        "_id": "Jobs_total",
        "name": "Total jobs",
      },
    ]
  }
  regionId="MOCK_REGION_ID"
>
  <Group
    label="Compare to"
  >
    <div
      className="form-group"
    >
      <label
        className="control-label"
        htmlFor="compare-to-input-3"
      >
        Compare to
      </label>
      <S
        getOptionLabel={[Function]}
        getOptionValue={[Function]}
        isClearable={true}
        onChange={[Function]}
        options={Array []}
      >
        <ReactSelect
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          isClearable={true}
          onChange={[Function]}
          options={Array []}
          styles={
            Object {
              "control": [Function],
              "option": [Function],
            }
          }
        >
          <div
            className="ReactSelectMock"
          />
        </ReactSelect>
      </S>
    </div>
  </Group>
  <Group
    label="Access to"
  >
    <div
      className="form-group"
    >
      <label
        className="control-label"
        htmlFor="access-to-input-4"
      >
        Access to
      </label>
      <P>
        <Box
          as="p"
          mb="10px"
        >
          <p
            className="css-1r2f04i"
          >
            Total jobs within 60 minutes (average instantaneous accessibility)
          </p>
        </Box>
      </P>
      <Legend
        breaks={
          Array [
            0,
            436,
            857,
            1257,
            1649,
            2000,
          ]
        }
        colors={
          Array [
            Object {
              "b": 255,
              "g": 255,
              "opacity": 0,
              "r": 255,
            },
            Rgb {
              "b": 239,
              "g": 219,
              "opacity": 0.42,
              "r": 198,
            },
            Rgb {
              "b": 225,
              "g": 202,
              "opacity": 0.42,
              "r": 158,
            },
            Rgb {
              "b": 214,
              "g": 174,
              "opacity": 0.42,
              "r": 107,
            },
            Rgb {
              "b": 189,
              "g": 130,
              "opacity": 0.42,
              "r": 49,
            },
            Rgb {
              "b": 156,
              "g": 81,
              "opacity": 0.42,
              "r": 8,
            },
          ]
        }
        min={0}
      >
        <div
          className="Legend"
        >
          <div
            className="Legend-Item"
            key="break-0"
            style={
              Object {
                "backgroundColor": "rgba(255, 255, 255, 1)",
                "color": "#000",
              }
            }
          >
            0 (transparent)
          </div>
          <div
            className="Legend-Item"
            key="break-1"
            style={
              Object {
                "backgroundColor": "rgba(198, 219, 239, 1)",
                "color": "#000",
              }
            }
          >
            1 to 436
          </div>
          <div
            className="Legend-Item"
            key="break-2"
            style={
              Object {
                "backgroundColor": "rgba(158, 202, 225, 1)",
                "color": "#000",
              }
            }
          >
            437 to 857
          </div>
          <div
            className="Legend-Item"
            key="break-3"
            style={
              Object {
                "backgroundColor": "rgba(107, 174, 214, 1)",
                "color": "#000",
              }
            }
          >
            858 to 1,257
          </div>
          <div
            className="Legend-Item"
            key="break-4"
            style={
              Object {
                "backgroundColor": "rgba(49, 130, 189, 1)",
                "color": "#fff",
              }
            }
          >
            1,258 to 1,649
          </div>
          <div
            className="Legend-Item"
            key="break-5"
            style={
              Object {
                "backgroundColor": "rgba(8, 81, 156, 1)",
                "color": "#fff",
              }
            }
          >
            1,650 to 2,000
          </div>
        </div>
      </Legend>
    </div>
  </Group>
  <AggregationArea
    regionId="MOCK_REGION_ID"
  >
    <Group
      label="Aggregate results to"
    >
      <div
        className="form-group"
      >
        <label
          className="control-label"
          htmlFor="aggregate-results-to-input-5"
        >
          Aggregate results to
        </label>
        <S
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          isClearable={true}
          isDisabled={false}
          name="aggregateTo"
          onChange={[Function]}
          options={Array []}
        >
          <ReactSelect
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            isClearable={true}
            isDisabled={false}
            name="aggregateTo"
            onChange={[Function]}
            options={Array []}
            styles={
              Object {
                "control": [Function],
                "option": [Function],
              }
            }
          >
            <div
              className="ReactSelectMock"
            />
          </ReactSelect>
        </S>
      </div>
    </Group>
    <Group>
      <div
        className="form-group"
      >
        <Button
          block={true}
          onClick={[Function]}
          size="sm"
          style="info"
        >
          <a
            className="btn btn-info btn-block btn-sm"
            href="#"
            onClick={[Function]}
            tabIndex={0}
          >
            Upload new aggregation area
             
            <Icon
              icon={
                Object {
                  "icon": Array [],
                  "iconName": "chevron-down",
                  "prefix": "fas",
                }
              }
            >
              <FontAwesomeIcon
                border={false}
                className=""
                fixedWidth={true}
                flip={null}
                icon={
                  Object {
                    "icon": Array [],
                    "iconName": "chevron-down",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                swapOpacity={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-chevron-down fa-w-NaN fa-fw "
                  data-icon="chevron-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 undefined undefined"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </Icon>
          </a>
        </Button>
      </div>
    </Group>
  </AggregationArea>
</RegionalResults>
`;
